core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpServiceMethod.java

```diff
@@ -161,65 +163,79 @@ private Object invoke(ServiceRequestContext ctx, HttpRequest req, @Nullable Aggr
-                        try {
-                            type = (Class<?>) ((ParameterizedType) p.getParameterizedType())
-                                    .getActualTypeArguments()[0];
-                        } catch (Exception e) {
-                            throw new IllegalArgumentException("Invalid optional parameter: " + p.getName(), e);
-                        }
-                    } else {
-                        type = p.getType();
-                    }
+                entries.add(createParamOrHeaderParam(parameterInfo, ParameterType.PARAM,
+                                                     pathParams, param.value()));
+                continue;
+            }
 
-                    entries.add(Parameter.ofParam(isMandatory, isOptionalType,
-                                                  validateAndNormalizeSupportedType(type), param.value(),
-                                                  defaultValue));
-                }
+            final Header header = parameterInfo.getAnnotation(Header.class);
+            if (header != null) {
+                entries.add(createParamOrHeaderParam(parameterInfo, ParameterType.HEADER,
+                                                     null, header.value()));
                 continue;
             }
 
-            if (p.getType().isAssignableFrom(ServiceRequestContext.class) ||
-                p.getType().isAssignableFrom(HttpParameters.class)) {
-                entries.add(Parameter.ofPredefinedType(p.getType()));
+            if (parameterInfo.getType().isAssignableFrom(ServiceRequestContext.class) ||
+                parameterInfo.getType().isAssignableFrom(HttpParameters.class)) {
+                entries.add(Parameter.ofPredefinedType(parameterInfo.getType()));
                 continue;
             }
 
-            if (p.getType().isAssignableFrom(HttpRequest.class) ||
-                p.getType().isAssignableFrom(AggregatedHttpMessage.class)) {
+            if (parameterInfo.getType().isAssignableFrom(HttpRequest.class) ||
+                parameterInfo.getType().isAssignableFrom(AggregatedHttpMessage.class)) {
                 if (hasRequestMessage) {
                     throw new IllegalArgumentException("Only one request message variable is allowed.");
                 }
                 hasRequestMessage = true;
-                entries.add(Parameter.ofPredefinedType(p.getType()));
+                entries.add(Parameter.ofPredefinedType(parameterInfo.getType()));
                 continue;
             }
 
-            throw new IllegalArgumentException("Unsupported object type: " + p.getType());
+            throw new IllegalArgumentException("Unsupported object type: " + parameterInfo.getType());
         }
         return entries.build();
     }
 
+    private static Parameter createParamOrHeaderParam(java.lang.reflect.Parameter parameterInfo,
+                                                      ParameterType paramType, final Set<String> pathParams,
```

Comment:
nit: As far as I know, we don't usually put `final` keyword on a method parameter.